### PR TITLE
feat: Add paypal

### LINKED ISSUE #333: add paypal

### PR DESCRIPTION
**Pull request by anhanh11001**
Fixes #333
Fixes #1813

**Screenshots**

**Additional Context**
Waiting for https://github.com/fossasia/open-event-server/pull/6297 to be merged, another PR will be made and open to send verification to the server before confirming orders

### CODE REVIEW

**Comment by liveHarshit** [comment]
PR is the same as #1813 What about the server-side confirmation?

**Comment by anhanh11001** [comment]
@liveHarshit still looking for documentation, this is an example for server-side integration but it is not fully enough for me: https://gist.github.com/tomwhipple/62075b4eb9abca1129cf

**Comment by liveHarshit** [comment]
> @liveHarshit still looking for documentation, this is an example for server-side integration but it is not fully enough for me: https://gist.github.com/tomwhipple/62075b4eb9abca1129cf
> 
> >  Gist**[Verify a PayPal immediate payment in sandbox](https://gist.github.com/tomwhipple/62075b4eb9abca1129cf)**Verify a PayPal immediate payment in sandbox. GitHub Gist: instantly share code, notes, and snippets.

Here I can see that - 
```
"payer": {
        "payment_method": "paypal", 
        "status": "VERIFIED", 
        "payer_info": {
            "first_name": "tom", 
            "last_name": "whipple", 
            "email": "twhipple-consumer@paypal.com", 
            "payer_id": "CBQG8NAHWQ3GE"
        }
    }, 
``` 
payer id is from server side response. But we want is to charge the payment according to our server implementation.

**Comment by anhanh11001** [comment]
Actually I just found a better documentation. We'll handle all the payment in the client side by the SDK and then only send the payment confirmation to the server, the server will check whether payment is confirmed or not like here: https://github.com/paypal/PayPal-Python-SDK/tree/master/samples/mobile_backend

**Comment by liveHarshit** [comment]
> send the payment confirmation to the server,

What information do we need to send on the server?

**Comment by anhanh11001** [comment]
only the payment id and then the server will search for that payment based on the payment ID. After that, server will check whether payment is already successful, if it is then it will send OK response to the Android so that Android app can proceed finishing order

**Comment by anhanh11001** [comment]
I think verify_payment function in here https://github.com/paypal/PayPal-Python-SDK/blob/master/samples/mobile_backend/paypal_client.py#L37 is what I have been looking for

**Comment by liveHarshit** [comment]
> only the payment id and then the server will search for that payment based on the payment ID. After that, server will check whether payment is already successful, if it is then it will send OK response to the Android so that Android app can proceed finishing order

Oh, that's nice. So for this server-side configuration is required. :v:

**Comment by liveHarshit** [comment]
> I think verify_payment function in here https://github.com/paypal/PayPal-Python-SDK/blob/master/samples/mobile_backend/paypal_client.py#L37 is what I have been looking for
> 
> >  GitHub**[paypal/PayPal-Python-SDK](https://github.com/paypal/PayPal-Python-SDK)**Python SDK for PayPal RESTful APIs. Contribute to paypal/PayPal-Python-SDK development by creating an account on GitHub.

Okay, so this configuration is required for issue https://github.com/fossasia/open-event-server/issues/5961

**Comment by liveHarshit** [comment]
Let's wait for the server PR to be merged, so that I can also test it.

File: app/build.gradle [hunk 1]
```diff
@@ -11,6 +11,7 @@ apply plugin: "com.github.b3er.local.properties"
 
 def STRIPE_API_TOKEN = System.getenv('STRIPE_API_TOKEN') ?: "YOUR_API_KEY"
 def MAPBOX_KEY = System.getenv('MAPBOX_KEY') ?: "pk.eyJ1IjoiYW5nbWFzMSIsImEiOiJjanNqZDd0N2YxN2Q5NDNuNTBiaGt6eHZqIn0.BCrxjW6rP_OuOuGtbhVEQg"
+def PAYPAL_API_TOKEN = System.getenv('PAYPAL_API_TOKEN') ?: "YOUR_API_KEY"
```
Comment (liveHarshit) [inline app/build.gradle]: I think it should be the client id.

**Comment by liveHarshit** [comment]
> **Screenshots**
> 

Is it with the sandbox or live?

**Comment by anhanh11001** [comment]
sandbox

File: app/build.gradle (hunk 1 above)
Comment (anhanh11001) [reply app/build.gradle]: yeah that's true, I'll change that

**Comment by liveHarshit** [comment]
@iamareebjamal Can I get PayPal live client id to test it?

**Comment by iamareebjamal** [comment]
Test with sandbox first.

**Comment by iamareebjamal** [comment]
Live won't even work as it is not gone to production

**Comment by liveHarshit** [comment]
> Test with sandbox first.

Then it needs to enable on dev instance with the same sandbox client id.

**Comment by anhanh11001** [comment]
@iamareebjamal I got 404 HTTP error for https://open-event-api-dev.herokuapp.com/v1/orders/e185e6d1-0d63-4a03-8dad-440fe6b8063d/verify-mobile-paypal-payment . Before that, the server PR was working fine for me on Postman. It is already in the debug server right?

**Comment by anhanh11001** [comment]
@liveHarshit You can try client ID: "AXrRge-gN1E5WHMYsu0FsftPOWBjTu3Sqj0pKfpq7mPKwv6w_ikiqWrka_06lAWZW5_WxL9CwaMiTVsc"

and Paypal email is letrananhduc0311-facilitator@gmail.com

**Comment by liveHarshit** [comment]
> @iamareebjamal I got 404 HTTP error for https://open-event-api-dev.herokuapp.com/v1/orders/e185e6d1-0d63-4a03-8dad-440fe6b8063d/verify-mobile-paypal-payment . Before that, the server PR was working fine locally for me on Postman. It is already in the debug server right?
> 
> > **[404 Not Found](https://open-event-api-dev.herokuapp.com/v1/orders/e185e6d1-0d63-4a03-8dad-440fe6b8063d/verify-mobile-paypal-payment)**

But it needs to unable on the server with the same client id, i think.

**Comment by iamareebjamal** [comment]
Try now @anhanh11001

**Comment by anhanh11001** [comment]
@iamareebjamal Got this error 
` paypalrestsdk.exceptions.MissingConfig: Required PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET.` Please enable paypal client ID and secret in debug server

**Comment by iamareebjamal** [comment]
Give the keys to me in private channel

**Comment by liveHarshit** [comment]
> Give the keys to me in private channel or you want keys from FOSSASIA sandbox?

We can use

> AXrRge-gN1E5WHMYsu0FsftPOWBjTu3Sqj0pKfpq7mPKwv6w_ikiqWrka_06lAWZW5_WxL9CwaMiTVsc

**Comment by iamareebjamal** [comment]
Give me the client secret for that in private chat

**Comment by liveHarshit** [comment]
> Paypal email is [letrananhduc0311-facilitator@gmail.com](mailto:letrananhduc0311-facilitator@gmail.com)

Also please share its password with me.

**Comment by liveHarshit** [comment]
> Give me the client secret for that in private chat

Please enable PayPal for debug instance

**Comment by iamareebjamal** [comment]
I asked a question to Duc about sandbox or live 2 days ago. Got the reply today

**Comment by iamareebjamal** [comment]
Added

**Comment by anhanh11001** [comment]
I still have the same problem of `paypalrestsdk.exceptions.MissingConfig: Required PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET`. Can you check whether it is deployed or not ?

**Comment by iamareebjamal** [comment]
Yes, it is set up in sandbox mode. If you want, I can change it to live mode but I doubt it'll work

**Comment by anhanh11001** [comment]
Here is the log
```
D/OkHttp: <-- 500 INTERNAL SERVER ERROR https://open-event-api-dev.herokuapp.com/v1/orders/ebb8edff-e67b-4dc6-b377-9ec33e2bcc65/verify-mobile-paypal-payment (322ms)
    Connection: close
    Content-Type: text/html; charset=utf-8
    X-Xss-Protection: 0
    Server: Werkzeug/0.15.4 Python/3.6.5
    Date: Wed, 21 Aug 2019 05:03:21 GMT
    Via: 1.1 vegur
```

```
paypalrestsdk.exceptions.MissingConfig
paypalrestsdk.exceptions.MissingConfig: Required PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET. Refer https://github.com/paypal/rest-api-sdk-python#configuration

Traceback (most recent call last)
File "/app/.heroku/python/lib/python3.6/os.py", line 666, in __getitem__
        self.unsetenv = unsetenv
        self._data = data
 
    def __getitem__(self, key):
        try:
            value = self._data[self.encodekey(key)]
        except KeyError:
            # raise KeyError with the original key value
            raise KeyError(key) from None
        return self.decodevalue(value)
 
During handling of the above exception, another exception occurred:
File "/app/.heroku/python/lib/python3.6/site-packages/paypalrestsdk/api.py", line 314, in default
    By default points to developer sandbox
    """
    global __api__
    if __api__ is None:
        try:
            client_id = os.environ["PAYPAL_CLIENT_ID"]
            client_secret = os.environ["PAYPAL_CLIENT_SECRET"]
        except KeyError:
            raise exceptions.MissingConfig("Required PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET. \
                Refer https://github.com/paypal/rest-api-sdk-python#configuration")
 
D/OkHttp:
File "/app/.heroku/python/lib/python3.6/os.py", line 669, in __getitem__
    def __getitem__(self, key):
        try:
            value = self._data[self.encodekey(key)]
        except KeyError:
            # raise KeyError with the original key value
            raise KeyError(key) from None
        return self.decodevalue(value)
 
    def __setitem__(self, key, value):
        key = self.encodekey(key)
        value = self.encodevalue(value)
During handling of the above exception, another exception occurred:
File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/flask.py", line 69, in sentry_patched_wsgi_app
            # type: (Any, Dict[str, str], Callable) -> _ScopedResponse
            if Hub.current.get_integration(FlaskIntegration) is None:
                return old_app(self, environ, start_response)
 
            return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
                environ, start_response
            )
 
        Flask.__call__ = sentry_patched_wsgi_app  # type: ignore
 
 
File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/wsgi.py", line 106, in __call__
                    rv = self.app(
                        environ,
                        functools.partial(_sentry_start_response, start_response, span),
                    )
                except BaseException:
                    reraise(*_capture_exception(hub))
 
        return _ScopedResponse(hub, rv)
 
 
def _sentry_start_response(
File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/_compat.py", D/OkHttp: line 54, in reraise
    def reraise(tp, value, tb=None):
        # type: (Optional[Type[BaseException]], Optional[BaseException], Optional[Any]) -> None
        assert value is not None
        if value.__traceback__ is not tb:
            raise value.with_traceback(tb)
        raise value
 
 
def with_metaclass(meta, *bases):
    class metaclass(type):
        def __new__(cls, name, this_bases, d):
File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/wsgi.py", line 103, in __call__
 
            with hub.start_span(span) as span:
                try:
                    rv = self.app(
                        environ,
                        functools.partial(_sentry_start_response, start_response, span),
                    )
                except BaseException:
                    reraise(*_capture_exception(hub))
 
        return _ScopedResponse(hub, rv)
File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/flask.py", line 68, in <lambda>
        def sentry_patched_wsgi_app(self, environ, start_response):
            # type: (Any, Dict[str, str], Callable) -> _ScopedResponse
            if Hub.current.get_integration(FlaskIntegration) is None:
                return old_app(self, environ, start_response)
 
            return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
                environ, start_response
            )
 
        Flask.__call__ = sentry_patched_wsgi_app  # type: ignore
 
File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 2463, D/OkHttp: in __call__
 
    def __call__(self, environ, start_response):
        """The WSGI server calls the Flask application object as the
        WSGI application. This calls :meth:`wsgi_app` which can be
        wrapped to applying middleware."""
        return self.wsgi_app(environ, start_response)
 
    def __repr__(self):
        return "<%s %r>" % (self.__class__.__name__, self.name)
File "/app/app/__init__.py", line 72, in __call__
        scheme = environ.get('HTTP_X_FORWARDED_PROTO')
        if scheme:
            environ['wsgi.url_scheme'] = scheme
        if os.getenv('FORCE_SSL', 'no') == 'yes':
            environ['wsgi.url_scheme'] = 'https'
        return self.app(environ, start_response)
 
 
app.wsgi_app = ReverseProxied(app.wsgi_app)
 
app_created = False
File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
            try:
                ctx.push()
                response = self.full_dispatch_request()
            except Exception as e:
                error = e
                response = self.handle_exception(e)
            except:  # noqa: B001
                error = sys.exc_info()[1]
                raise
            return response(environ, start_response)
        finally:
File "/app/.heroku/python/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
        # Wrap exception handlers with cross_origin
        # These error handlers will still respect the behavior of the route
        if options.get('intercept_exceptions', True):
            def _after_request_decorator(f):
                def wrapped_function(*args, **kwargs):
D/OkHttp:
                    return cors_after_request(app.make_response(f(*args, **kwargs)))
                return wrapped_function
 
            if hasattr(app, 'handle_exception'):
                app.handle_exception = _after_request_decorator(
                    app.handle_exception)
File "/app/.heroku/python/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
        # Wrap exception handlers with cross_origin
        # These error handlers will still respect the behavior of the route
        if options.get('intercept_exceptions', True):
            def _after_request_decorator(f):
                def wrapped_function(*args, **kwargs):
                    return cors_after_request(app.make_response(f(*args, **kwargs)))
                return wrapped_function
 
            if hasattr(app, 'handle_exception'):
                app.handle_exception = _after_request_decorator(
                    app.handle_exception)
File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1866, in handle_exception
            # if we want to repropagate the exception, we can attempt to
            # raise it with the whole traceback in case we can do that
            # (the function was actually called from the except part)
            # otherwise, we just raise the error again
            if exc_value is e:
                reraise(exc_type, exc_value, tb)
            else:
                raise e
 
        self.log_exception((exc_type, exc_value, tb))
        server_error = InternalServerError()
File "/app/.heroku/python/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    import collections.abc as collections_abc
 
    def reraise(tp, value, tb=None):
D/OkHttp:
 
D/OkHttp:
        return self.view_functions[rule.endpoint](**req.view_args)
 
    def full_dispatch_request(self):
        """Dispatches the request and on top of that performs request
        pre and postprocessing as well as HTTP exception catching and
        error handling.
    @wraps(fn)
        verify_jwt_in_request()
        current_user.last_accessed_at = datetime.utcnow()
        save_to_db(current_user)
        return fn(*args, **kwargs)
D/OkHttp:
File "/app/app/api/helpers/payment.py", line 213, in verify_payment
    def verify_payment(payment_id, order):
KeyError: 'PAYPAL_CLIENT_ID'
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/flask.py", line 69, in sentry_patched_wsgi_app
        environ, start_response
      File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/wsgi.py", line 106, in __call__
        reraise(*_capture_exception(hub))
      File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/_compat.py", line 54, in reraise
        raise value
      File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/wsgi.py", line 103, in __call__
        functools.partial(_sentry_start_response, start_response, span),
      File "/app/.heroku/python/lib/python3.6/site-packages/sentry_sdk/integrations/flask.py", line 68, in <lambda>
        return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
      File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 2463, in __call__
        return self.wsgi_app(environ, start_response)
      File "/app/app/__init__.py", line 72, in __call__
        return self.app(environ, start_response)
      File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
        response = self.handle_exception(e)
      File "/app/.heroku/python/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
        return cors_after_request(app.make_response(f(*args, **kwargs)))
```

**Comment by iamareebjamal** [comment]
I have changed it to live credentials. Test now. If not working, please raise issue on server

**Comment by anhanh11001** [comment]
Yes, it is not working as well and doesn't provide any logs. What should be the problem raised in the server?

**Comment by iamareebjamal** [comment]
https://github.com/fossasia/open-event-attendee-android/pull/2223#issuecomment-523298727

**Comment by iamareebjamal** [comment]
Does not give any log? What's the response?

**Comment by iamareebjamal** [comment]
I have restarted the server. Try one time more

**Comment by anhanh11001** [comment]
oh I think I got the problem, I didn't put the function to configPaypal() in the verifyPayment() method. But when I test, I didn't know where to put the CLIENT_ID and SECRET so I put it directly to the path where the path logs direct me, that's probably why

**Comment by liveHarshit** [comment]
Is it working now?

**Comment by anhanh11001** [comment]
@liveHarshit still having some small mistake, sorry, it is so annoying, I forgot to push the change

**Comment by anhanh11001** [comment]
In Order, when formatting to Json, isBillingEnabled turn into "billing_enabled", not "is_billing_enabled". How should I fix this?

**Comment by iamareebjamal** [comment]
Add @JsonPropert annotation

**Comment by liveHarshit** [comment]
> In Order, when formatting to Json, isBillingEnabled turn into "billing_enabled", not "is_billing_enabled". How should I fix this?

Use 1@get:JsonProperty("is-read")1

**Comment by anhanh11001** [comment]
tested and working for normal payment. This is just a general implementation, I will open more issues on different currency problems like billing address in US requires state location or Japanese,.. currency should not contain fractional part

**Comment by aggarwalpulkit596** [comment]
@anhanh11001 how to check PayPal integration in debug mode?

**Pull request merged**